File: frontend/src/types/documentUpload.ts
import type { DocumentFileStatus, SupportedDocumentExtension } from './documentFile';
import type { UploadSessionStatus, UploadSessionType } from './uploadSession';

export const fileIdentificationStatuses = [
  'IDENTIFIED',
  'PARTIALLY_IDENTIFIED',
  'NOT_IDENTIFIED',
  'DUPLICATE_FILE',
  'INVALID',
] as const;

export type FileIdentificationStatus = (typeof fileIdentificationStatuses)[number];

export const uploadProposedActions = [
  'ATTACH_TO_EXISTING_REVISION',
  'CREATE_DOCUMENT_AND_REVISION',
  'ADD_NEW_REVISION',
  'REPLACE_CURRENT_FILE',
  'MANUAL_REVIEW',
  'SKIP',
] as const;

export type UploadProposedAction = (typeof uploadProposedActions)[number];

export const uploadSessionItemStatuses = [
  'PENDING',
  'READY',
  'COMMITTED',
  'SKIPPED',
  'FAILED',
  'CANCELLED',
] as const;

export type UploadSessionItemStatus = (typeof uploadSessionItemStatuses)[number];

export interface ParsedDocumentMetadata {
  companyCode?: string | null;
  departmentCode?: string | null;
  sectionCode?: string | null;
  documentTypeCode?: string | null;
  documentNumber?: string | null;
  revisionCode?: string | null;
  baseDocumentCode?: string | null;
  fullDocumentCode?: string | null;
}

export interface MatchedUploadDocument {
  id: string;
  baseDocumentCode: string;
  title: string;
}

export interface MatchedUploadRevision {
  id: string;
  revisionCode: string;
  fullDocumentCode?: string | null;
}

export interface UploadSessionItem {
  uploadItemId: string;
  originalFilename: string;
  sanitizedFilename: string;
  fileExtension: SupportedDocumentExtension | null;
  mimeType: string | null;
  detectedMimeType: string | null;
  fileSize: number | null;
  sha256Hash: string | null;
  identificationStatus: FileIdentificationStatus;
  proposedAction: UploadProposedAction;
  parsedMetadata: ParsedDocumentMetadata | null;
  matchedDocument: MatchedUploadDocument | null;
  matchedRevision: MatchedUploadRevision | null;
  duplicateWarning?: FileDuplicateWarning | null;
  warnings: string[];
  errors: string[];
  status: UploadSessionItemStatus;
  quarantineReason?: string | null;
}

export interface UploadPreviewResponse {
  sessionId: string;
  sessionType: UploadSessionType;
  status: UploadSessionStatus;
  totalFiles: number;
  totalSize: number;
  expiresAt: string;
  committedAt: string | null;
  cancelledAt: string | null;
  items: UploadSessionItem[];
}

export interface FileDuplicateWarning {
  message: string;
  sameRevision: boolean;
  documentId?: string | null;
  revisionId?: string | null;
  baseDocumentCode?: string | null;
}

export interface UploadConfirmationMetadata {
  companyCode?: string | null;
  departmentId?: string | null;
  sectionId?: string | null;
  documentTypeId?: string | null;
  documentNumber?: string | null;
  title?: string | null;
  description?: string | null;
  revisionCode?: string | null;
  documentStatusId?: string | null;
  validationRuleId?: string | null;
  issueDate?: string | null;
  effectiveDate?: string | null;
  reviewDate?: string | null;
  expiryDate?: string | null;
  sharepointUrl?: string | null;
  externalReference?: string | null;
  remarks?: string | null;
  setAsCurrentRevision?: boolean;
  reason?: string | null;
  allowDuplicate?: boolean;
}

export interface UploadConfirmationItem {
  uploadItemId: string;
  action: UploadProposedAction;
  documentId?: string | null;
  revisionId?: string | null;
  metadata?: UploadConfirmationMetadata | null;
}

export interface UploadConfirmationRequest {
  items: UploadConfirmationItem[];
}

export interface UploadConfirmationItemResult {
  uploadItemId: string;
  action: UploadProposedAction;
  status: UploadSessionItemStatus;
  documentId: string | null;
  revisionId: string | null;
  documentFileId: string | null;
  baseDocumentCode: string | null;
  revisionCode: string | null;
  fileStatus: DocumentFileStatus | null;
  error: string | null;
}

export interface UploadConfirmationResult {
  sessionId: string;
  status: UploadSessionStatus;
  items: UploadConfirmationItemResult[];
  total?: number;
  committed?: number;
  skipped?: number;
  failed?: number;
  documentsCreated?: number;
  revisionsCreated?: number;
  filesAttached?: number;
  filesReplaced?: number;
  committedAt?: string | null;
}

export type UploadProgressHandler = (progress: number) => void;

File: backend/app/schemas/document_upload.py
"""Schemas for upload preview, identification, and confirmation."""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import Field, HttpUrl, model_validator

from app.models.upload_session_item import (
    UploadIdentificationStatus,
    UploadProposedAction,
    UploadSessionItemStatus,
)
from app.schemas.base import ApiSchema


class ParsedDocumentMetadata(ApiSchema):
    company_code: str | None = None
    department_code: str | None = None
    section_code: str | None = None
    document_type_code: str | None = None
    document_number: str | None = None
    title: str | None = Field(default=None, max_length=500)
    revision_code: str | None = None
    base_document_code: str | None = None
    full_document_code: str | None = None


class MatchedDocumentReference(ApiSchema):
    id: UUID
    base_document_code: str
    title: str


class MatchedRevisionReference(ApiSchema):
    id: UUID
    revision_code: str
    full_document_code: str


class FileDuplicateWarning(ApiSchema):
    message: str = "Duplicate file already exists."
    same_revision: bool = False
    document_id: UUID | None = None
    revision_id: UUID | None = None
    base_document_code: str | None = None


class FileIdentificationResult(ApiSchema):
    upload_item_id: UUID
    original_filename: str
    sanitized_filename: str
    file_extension: str | None = None
    mime_type: str | None = None
    detected_mime_type: str | None = None
    file_size: int | None = Field(default=None, ge=0)
    sha256_hash: str | None = None
    identification_status: UploadIdentificationStatus
    proposed_action: UploadProposedAction
    parsed_metadata: ParsedDocumentMetadata | None = None
    matched_document: MatchedDocumentReference | None = None
    matched_revision: MatchedRevisionReference | None = None
    duplicate_warning: FileDuplicateWarning | None = None
    warnings: list[str] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)
    quarantine_reason: str | None = None
    status: UploadSessionItemStatus


class UploadActionMetadata(ApiSchema):
    """Union-shaped metadata validated again for the selected action."""

    company_code: str | None = Field(default=None, max_length=20)
    department_id: UUID | None = None
    section_id: UUID | None = None
    document_type_id: UUID | None = None
    document_number: str | None = Field(default=None, max_length=50)
    title: str | None = Field(default=None, max_length=500)
    description: str | None = None
    revision_code: str | None = Field(default=None, max_length=30)
    document_status_id: UUID | None = None
    validation_rule_id: UUID | None = None
    issue_date: date | None = None
    effective_date: date | None = None
    review_date: date | None = None
    expiry_date: date | None = None
    sharepoint_url: HttpUrl | None = None
    external_reference: str | None = Field(default=None, max_length=500)
    remarks: str | None = None
    set_as_current_revision: bool = True
    reason: str | None = Field(default=None, max_length=1000)
    allow_duplicate: bool = False


class UploadConfirmationItem(ApiSchema):
    upload_item_id: UUID
    action: UploadProposedAction
    document_id: UUID | None = None
    revision_id: UUID | None = None
    metadata: UploadActionMetadata | None = None

    @model_validator(mode="after")
    def validate_action_targets(self) -> UploadConfirmationItem:
        if self.action == UploadProposedAction.ATTACH_TO_EXISTING_REVISION:
            if self.document_id is None or self.revision_id is None:
                raise ValueError(
                    "documentId and revisionId are required when attaching "
                    "to an existing revision."
                )
        elif self.action == UploadProposedAction.CREATE_DOCUMENT_AND_REVISION:
            if self.metadata is None:
                raise ValueError(
                    "metadata is required when creating a document."
                )
            required = {
                "departmentId": self.metadata.department_id,
                "documentTypeId": self.metadata.document_type_id,
                "documentNumber": self.metadata.document_number,
                "title": self.metadata.title,
                "revisionCode": self.metadata.revision_code,
            }
            missing = [name for name, value in required.items() if not value]
            if missing:
                raise ValueError(
                    f"{', '.join(missing)} required when creating a document."
                )
        elif self.action == UploadProposedAction.ADD_NEW_REVISION:
            if self.document_id is None:
                raise ValueError(
                    "documentId is required when adding a revision."
                )
            if self.metadata is None or not self.metadata.revision_code:
                raise ValueError(
                    "metadata.revisionCode is required when adding a revision."
                )
        elif self.action == UploadProposedAction.REPLACE_CURRENT_FILE:
            if self.document_id is None or self.revision_id is None:
                raise ValueError(
                    "documentId and revisionId are required when replacing a "
                    "file."
                )
            if self.metadata is None or not (
                self.metadata.reason and self.metadata.reason.strip()
            ):
                raise ValueError(
                    "metadata.reason is required when replacing a file."
                )
        return self


class UploadConfirmationRequest(ApiSchema):
    items: list[UploadConfirmationItem] = Field(
        min_length=1,
        max_length=500,
    )


class UploadConfirmationItemResult(ApiSchema):
    upload_item_id: UUID
    action: UploadProposedAction
    status: UploadSessionItemStatus
    document_id: UUID | None = None
    revision_id: UUID | None = None
    document_file_id: UUID | None = None
    base_document_code: str | None = None
    revision_code: str | None = None
    file_status: str | None = None
    error: str | None = None


class UploadConfirmationResult(ApiSchema):
    session_id: UUID
    status: str
    items: list[UploadConfirmationItemResult]


class BatchUploadConfirmationRequest(UploadConfirmationRequest):
    """Batch confirmation has the same item contract."""


class BatchUploadResult(ApiSchema):
    session_id: UUID
    status: str
    total: int = Field(ge=0)
    committed: int = Field(ge=0)
    skipped: int = Field(ge=0)
    failed: int = Field(ge=0)
    documents_created: int = Field(ge=0)
    revisions_created: int = Field(ge=0)
    files_attached: int = Field(ge=0)
    files_replaced: int = Field(ge=0)
    items: list[UploadConfirmationItemResult]
    committed_at: datetime | None = None
